src/h2olog/h2olog.cc

```diff
@@ -226,52 +225,31 @@ static void lost_cb(void *context, uint64_t lost)
     tracer->handle_lost(lost);
 }
 
-static size_t add_matched_usdts(std::vector<h2o_tracer::usdt> &selected_usdts, const std::vector<h2o_tracer::usdt> &available_usdts,
-                                const char *pattern)
-{
-    size_t added = 0;
-    for (const auto &usdt : available_usdts) {
-        if (fnmatch(pattern, usdt.fully_qualified_name().c_str(), 0) == 0) {
-            selected_usdts.push_back(usdt);
-            added++;
-        }
-    }
-    return added;
-}
-
 int main(int argc, char **argv)
 {
-    std::unique_ptr<h2o_tracer> tracer;
-    if (argc > 1 && strcmp(argv[1], "quic") == 0) {
-        fprintf(stderr, "warning: 'quic' mode is deprecated. Use 'raw' mode instead.\n");
-        tracer.reset(create_raw_tracer());
-        --argc;
-        ++argv;
-    } else if (argc > 1 && strcmp(argv[1], "raw") == 0) {
-        tracer.reset(create_raw_tracer());
-        --argc;
-        ++argv;
-    } else {
-        tracer.reset(create_http_tracer());
-    }
-
-    const std::vector<h2o_tracer::usdt> available_usdts = tracer->usdt_probes();
+    std::unique_ptr<h2o_tracer> tracer(create_raw_tracer());
 
     int debug = 0;
-    int preserve_root = 0;
+    bool preserve_root = 0;
+    bool list_usdts = false;
     FILE *outfp = stdout;
-    std::vector<h2o_tracer::usdt> selected_usdts;
     std::vector<std::string> response_header_filters;
     int c;
     pid_t h2o_pid = -1;
-    while ((c = getopt(argc, argv, "hdrlp:t:s:w:")) != -1) {
+    while ((c = getopt(argc, argv, "hHdrlp:t:s:w:")) != -1) {
         switch (c) {
+        case 'H':
+            tracer.reset(create_http_tracer());
+            break;
         case 'p':
             h2o_pid = atoi(optarg);
             break;
         case 't': {
-            size_t added = add_matched_usdts(selected_usdts, available_usdts, optarg);
-            if (added == 0) {
+            if (!tracer->can_select_usdts()) {
+                fprintf(stderr, "-t is not suppored in %s\n", tracer->name().c_str());
+                exit(EXIT_FAILURE);
+            }
+            if (!tracer->select_usdts(optarg)) {
```

Comment:
I might suggest doing the following:
* Change the return type o `select_usdts()` to `std::string` where the value is the error message (or an empty string if successful).
* Let the h2o_http_tracer::select_usdts() emit "-t is not supported ...".
* Remove `can_select_usdts` and `name` methods.

Reply:
Thank you for your review. It makes sense. 

3eec446db

> 4 files changed, 16 insertions(+), 36 deletions(-)

Much more simplified!